### PR TITLE
Avoid bigint_cast stack overflow

### PR DESCRIPTION
**Pull request by ucasfl**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid possible stack overflow in bigint conversion. Big integers are experimental.

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: base/common/wide_integer_impl.h [hunk 1]
```diff
@@ -239,6 +239,13 @@ struct integer<Bits, Signed>::_impl
     template <class T>
     constexpr static void set_multiplier(integer<Bits, Signed> & self, T t) noexcept {
         constexpr uint64_t max_int = std::numeric_limits<uint64_t>::max();
+
+        /// Replacing nan to zero, and inf to max(T) to avoid stack overflow in bigint_cast
+        if (isnan(t))
+            t = 0;
+        if (isinf(t))
```
Comment (alexey-milovidov) [inline base/common/wide_integer_impl.h:246]: There is also -inf that probably better to be replaced to the min value.
But as it's implementation specific behaviour, let's replace all to zero.

**Review by alexey-milovidov — APPROVED** [review]

File: base/common/wide_integer.h [hunk 2]
```diff
@@ -127,7 +127,7 @@ class integer
     friend class std::numeric_limits<integer<Bits, signed>>;
     friend class std::numeric_limits<integer<Bits, unsigned>>;
 
-    base_type items[_impl::item_count];
+    base_type items[_impl::item_count] = {};
```
Comment (alexey-milovidov) [inline base/common/wide_integer.h:130]: No, we should not have it always initialized. It should behave in the same way as builtin types.

Comment (ucasFL) [reply base/common/wide_integer.h]: Ok, seems that is a gcc bug.

**Comment by ucasFL** [comment]
Build failed on gcc-9: `constructor is not user-provided because it is explicitly defaulted in the class body, ... and the implicitly-defined constructor does not initialize 'wide::integer<256, unsigned int>::base_type wide::integer<256, unsigned int>::items [4]`.

Seems this is a gcc bug [#80935](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80935), and I don't know how to prevent it.

**Pull request merged**